plugin/src/main/java/org/autorefactor/refactoring/rules/WakeLockRefactoring.java

```diff
@@ -0,0 +1,208 @@
+            final ASTBuilder b = this.ctx.getASTBuilder();
+            MethodDeclaration enclosingMethod;
+            enclosingMethod = (MethodDeclaration) ASTNodes.getParent(node, ASTNode.METHOD_DECLARATION);
+            if (isMethod(enclosingMethod.resolveBinding(), "android.app.Activity", "onDestroy")) {
+                TypeDeclaration typeDeclaration;
+                typeDeclaration = (TypeDeclaration) ASTNodes.getParent(enclosingMethod, TypeDeclaration.class);
+                MethodDeclaration onPauseMethod = findMethodOfType("onPause", typeDeclaration);
+                if (onPauseMethod != null && node.getParent().getNodeType() == ASTNode.EXPRESSION_STATEMENT) {
+                    Statement releaseNode = createWakelockReleaseNode(node);
+                    r.remove(node.getParent());
+                    r.insertAt(releaseNode,
+                            onPauseMethod.getBody().statements().size(),
+                            Block.STATEMENTS_PROPERTY,
+                            onPauseMethod.getBody());
+                    return DO_NOT_VISIT_SUBTREE;
+                }
+                /* If it reaches this part of the code, it
+                 * means it did not find onPause method in the class.
+                 */
+                MethodDeclaration onPauseDeclaration = createOnPauseDeclaration();
+
+                // add onPause declaration to the Activity
+                r.insertAfter(onPauseDeclaration, enclosingMethod);
+                return DO_NOT_VISIT_SUBTREE;
+            }
+        } else if (isMethod(node, "android.os.PowerManager.WakeLock", "acquire")) {
+            final Refactorings r = this.ctx.getRefactorings();
+            final ASTBuilder b = this.ctx.getASTBuilder();
+            TypeDeclaration typeDeclaration = (TypeDeclaration) ASTNodes.getParent(node, ASTNode.TYPE_DECLARATION);
+            ReleasePresenceChecker releasePresenceChecker = new ReleasePresenceChecker();
+            typeDeclaration.accept(releasePresenceChecker);
+            if (!releasePresenceChecker.releasePresent) {
+                Statement releaseNode = createWakelockReleaseNode(node);
+                MethodDeclaration onPauseMethod = findMethodOfType("onPause", typeDeclaration);
+                if (onPauseMethod != null && node.getParent().getNodeType() == ASTNode.EXPRESSION_STATEMENT) {
+                    r.insertAt(releaseNode,
+                            onPauseMethod.getBody().statements().size(),
+                            Block.STATEMENTS_PROPERTY,
+                            onPauseMethod.getBody());
+                    return DO_NOT_VISIT_SUBTREE;
+                } else {
+                    MethodDeclaration onPauseDeclaration = createOnPauseDeclaration();
+                    /* TODO @JnRouvignac, instead of using insertAt, calling
+                     * typeDeclaration.bodyDeclarations().add(onPauseDeclaration);
+                     * would be more intuitive.
```

Comment:
Yes let's do that first and see how much it helps.
You can use http://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.jdt.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjdt%2Fcore%2Fdom%2Frewrite%2FListRewrite.html underneath